pkg/controllers/switchedcontroller/switched_controller.go

```diff
@@ -0,0 +1,127 @@
+// defined by the following:
+// - delegateFactoryFn: a function that is invoked when a new instance of the delegate controller must be created
+// - switchConditionFn: a function that is invoked on every switched controller sync to determine whether it needs to switch on/off the delegate controller
+// - informers: any informers that must be tracked and are required by the switch condition
+//
+// If no delegateFactoryFn is defined, the controller's sync will always fail. If no switchConditionFn is defined,
+// the delegate controller will never be switched off.
+func NewControllerWithSwitch(
+	operatorClient v1helpers.OperatorClient,
+	delegateName string,
+	delegateFactoryFn DelegateFactoryFunc,
+	switchConditionFn DelegateSwitchCondition,
+	informers []factory.Informer,
+	resyncInterval time.Duration,
+	eventRecorder events.Recorder,
+) factory.Controller {
+
+	c := &ControllerWithSwitch{
+		delegateName:      delegateName,
+		delegateFactoryFn: delegateFactoryFn,
+		switchConditionFn: switchConditionFn,
+	}
+
+	return factory.New().
+		WithSync(c.sync).
+		WithSyncDegradedOnError(operatorClient).
+		WithInformers(informers...).
+		ResyncEvery(resyncInterval).
+		ToController(delegateName+"_SwitchedController", eventRecorder)
+}
+
+func (c *ControllerWithSwitch) runDelegate(ctx context.Context, syncCtx factory.SyncContext) {
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+
+	if c.delegateController == nil {
+		c.delegateController = c.delegateFactoryFn(ctx).ToController(c.delegateName, syncCtx.Recorder())
+	}
+	go c.delegateController.Run(ctx, 1)
+}
+
+func (c *ControllerWithSwitch) stopDelegate() {
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+
+	if c.switchContextCancel != nil {
+		c.switchContextCancel()
+	}
+
+	c.switchContext = nil
+	c.delegateController = nil
+}
+
+func (c *ControllerWithSwitch) sync(ctx context.Context, syncCtx factory.SyncContext) error {
+	if c.delegateFactoryFn == nil {
+		return fmt.Errorf("no delegate factory function defined")
+	}
+
+	if c.switchConditionFn == nil {
+		klog.Infof("no switch condition defined; controller '%s' will never be stopped", c.delegateName)
+		c.runDelegate(ctx, syncCtx)
+		return nil
+	}
+
+	switchOn, err := c.switchConditionFn()
+	if err != nil {
+		return fmt.Errorf("could not determine switch condition: %v", err)
+	}
+
+	switch {
+	case switchOn && c.switchContext == nil:
+		c.switchContext, c.switchContextCancel = context.WithCancel(ctx)
+		c.runDelegate(c.switchContext, syncCtx)
+
+	case switchOn && c.switchContext != nil && c.switchContext.Err() == nil:
+		// context alive, delegate running
+
+	case !switchOn && c.switchContext == nil:
+		// we haven't been asked to start yet
+
+	case !switchOn && c.switchContext != nil && c.switchContext.Err() == nil:
+		c.stopDelegate()
+
+	default:
+		return fmt.Errorf("this should never happen; switchOn = %v; switchContext error = %v", switchOn, c.switchContext.Err())
```

Comment:
We would end up hitting this in the scenario where:
- `switchOn = true`
- `c.switchContext != nil`
- `c.switchContext.Err() != nil`

I.e - we want the controller to be running, but the context has been canceled and is therefore not running.

I think we would _also_ hit this statement in the scenario where we want to switch the controller off but it has already been switched off by a canceled context.

Because we are using a passed in context, are we actually able to guarantee that never happens?